packages/context/src/context.ts

```diff
@@ -222,7 +238,7 @@ export class Context {
    *   (deeply) nested property to retrieve.
    * @returns A promise of the bound value.
    */
-  get<T>(keyWithPath: string): Promise<T>;
+  get<T>(keyWithPath: string | BindingKey<T>): Promise<T>;
```

Comment:
Can we get rid of using generics for these functions? Instead of this `get<string>('some.key')`, can we force users to do this: `get(new BindingKey<string>('some.key'))`

Reply:
We could, but I don't want to. Think about people using JavaScript - I prefer `ctx.get('some.key')` over `ctx.get(new BindingKey('some.key'))`.

Reply:
People using JavaScript wouldn't get type coersion anyway, so they're free to choose either ways of retrieving their dependencies (I'd think that most would go for the `ctx.get('some.key')` option). What I don't like about the hard type casting at the function level is that it's something that users could get wrong and not get any errors from. But the same could be said about the change you're introducing, so maybe it's ok?

Reply:
Well, if you force users to do `get(new BindingKey<string>('some.key'))` then JavaScript users won't be able to write `ctx.get('some.key')`. Unless I misunderstood you?

> What I don't like about the hard type casting at the function level is that it's something that users could get wrong and not get any errors from. 

Could you please post a code snippet demonstrating such invalid usage? I like making APIs difficult to misuse, but I am having difficulties imagining the example you have in mind.

Reply:
My bad, I meant that we should force users to do this: `get(new BindingKey<string>('some.key')` OR `get('some.key')` (notice the lack of generics here). So if they want type coercion, we have them use the `BindingKey` constructor.

Take a look at this example:
```ts
const key = 'key';
const value = 'value';

const b = new Context();
b.bind(value).to(key);
const retrievedValue = b.getSync<number>(key);
```

There is going to be no compilation error here since there is no way of knowing if  `retrievedValue` is actually a number.
Since we're going to be exporting the keys in the form of `BindingKey` in most of our modules, if users were to use `get()` to retrieve whatever's bound to the BindingKey they're importing, they'd immediately know what was the type of the value gotten.